src/theory/datatypes/theory_datatypes_type_rules.h

```diff
@@ -427,7 +427,115 @@ class DtSyguEvalTypeRule
     }
     return TypeNode::fromType(dt.getSygusType());
   }
-}; /* class DtSygusBoundTypeRule */
+}; /* class DtSyguEvalTypeRule */
+
+class MatchTypeRule
+{
+ public:
+  inline static TypeNode computeType(NodeManager* nodeManager,
+                                     TNode n,
+                                     bool check)
+  {
+    Assert(n.getKind() == kind::MATCH);
+
+    TypeNode retType;
+
+    TypeNode headType = n[0].getType(check);
+    if (!headType.isDatatype())
+    {
+      throw TypeCheckingExceptionPrivate(n, "expecting datatype head in match");
+    }
+    const Datatype& hdt = headType.getDatatype();
+
+    // the type of a match case list is the least common type of its cases
+    for (unsigned i = 1, nchildren = n.getNumChildren(); i < nchildren; i++)
+    {
+      Node nc = n[i];
+      Kind nck = nc.getKind();
+      if (nck != kind::MATCH_CASE && nck != kind::MATCH_BIND_CASE)
+      {
+        throw TypeCheckingExceptionPrivate(
+            n, "expected a match case in match expression");
+      }
+      // get the pattern type
+      unsigned pindex = nck == kind::MATCH_CASE ? 0 : 1;
+      TypeNode patType = nc[pindex].getType();
+      // should be caught in the above call
+      Assert(patType.isDatatype());
+      const Datatype& pdt = patType.getDatatype();
+      // compare datatypes instead of the types to catch parametric case, where
+      // the pattern has parametric type.
+      if (hdt != pdt)
+      {
+        std::stringstream ss;
+        ss << "pattern of a match case does not match the head type in match";
+        throw TypeCheckingExceptionPrivate(n, ss.str());
+      }
+      TypeNode currType = nc.getType(check);
+      if (i == 1)
+      {
+        retType = currType;
+      }
+      else
+      {
+        retType = TypeNode::leastCommonTypeNode(retType, currType);
+        if (retType.isNull())
+        {
+          throw TypeCheckingExceptionPrivate(
+              n, "incomparable types in match case list");
+        }
+      }
+    }
+    return retType;
```

Comment:
We should probably also make sure that the default case only appears last, otherwise, we may be generating `ITE`s with null conditions in the rewriter.

Reply:
Right, I'll add this.

Reply:
I'll actually make the code a bit safer and make the case "true" for variables.  That way we could in theory support non-standard constructions of MATCH in case they slipped through somehow.